src/tqdm_publisher/_progress_handler.py

```diff
@@ -0,0 +1,27 @@
+import queue
+
+from ._progress_subscriber import TQDMProgressSubscriber
+
+
+class TQDMProgressHandler:
+    def __init__(self):
+        self.listeners = []
+
+    def listen(self):
+        q = queue.Queue(maxsize=25)
+        self.listeners.append(q)
+        return q
+
+    def create(self, iterable, additional_metadata: dict = dict(), **tqdm_kwargs):
```

Comment:
Is there a reason this has to be a separate novel method `create` instead of the `__init__`?

Reply:
In particular, it's a little odd to have to use this roundabout approach to 'create' a new one. You _might_ see this structure as a class constructor like

```python
@classmethod
def from_iterable(cls, iterable: Iterable, ...) -> Self:
    ....
```

but in general, inheriting from one of the parents and using OOP pattern would be preferred

Reply:
This class handles _more than one_ progress bar, so we can't just directly inherit from one of those classes.

This could instead be called `register` and receive an instantiated class from `tqdm_publisher`: 

```
progress_bar = TQDMProgressSubscriber(iterable)
progress_handler.register(progress_bar, additional_metadata=dict(request_id=request_id))
```

Where:

```
    def register(self, progress_bar, additional_metadata: dict = dict()):
        progress_bar.subscribe(lambda progress_update: self._announce(dict(
            **progress_update,
            **additional_metadata
        ))
```

It only doesn't work this way because it wouldn't be compatible with passing a `progress_bar_class` that is instantiated with `progress_bar_options`.

Reply:
We can add this as another method—but AFAIK we need the original for our downstream code.

Reply:
> This class handles more than one progress bar

Why do we need a single class to handle more than one bar?

The natural use of `tqdm` that we seek to replace is one progress bar per `iterable` we want to track progress on, we're just looking to reroute the normal `stdout` printouts to a frontend client. So is that not possible with $N$ progress handlers (one per $N$ iterables) replacing $N$ previous `tqdm` bars?

Reply:
Ah I'd written this down for myself but never noted it on the new PRs. We need a single class to handle **more than one bar's updates** and forward them to the client.

Server-side events requires that the connection is established on one endpoint (`listen_to_neuroconv_events`) and pass updates from other endpoints (e.g. `convert_to_nwb`, `inspect_nwb_file`, `upload_folder_to_dandi`, etc) here—usually in a queue-based instance like the `MessageAnnouncer`.

The new `TQDMProgressHandler` class organizes the messages from one or more progress bars into a queue, allowing these updates to conform to SSE's structural requirement and, in general, could be used to avoid overloading when running many parallel bars like in the WebSockets demo.

Reply:
Ah I was trying to communicate this here, things just got quite spread out: https://github.com/catalystneuro/tqdm_publisher/pull/46#discussion_r1532612489

Reply:
Aha. OK, I see now, thanks

I will think more on this then...